File: python/placerWidget.py
#-*- coding: utf-8 -*-
# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
from PySide6 import QtCore, QtGui, QtWidgets
from dataclasses import dataclass

# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
class ScaledLayout():
    def __init__(self,parent):  # parent required!
        self.arr = []
        geom = parent.geometry()
        self.oldParentSize = QtCore.QSizeF(geom.width(),geom.height())

    def deleteWidget(self,widget):
        for i in self.arr:
            if widget == i.widget:
                self.arr.remove(i) # remove from the array
                break
    
    def addWidget(self,widget):
        name = "unknown"
        try:
            if widget.objectName() != None:
                name = widget.objectName()
        except:
            print("widget has no object name")
        #
        # some children of widgets are not "widgets" and won't
        # have any geometry. The 'except' will take care of
        # that
        #
        try:
            geom = widget.geometry()
            font = widget.font()
            data = WidgetData(name,
                              widget,
                              geom.width(),
                              geom.height(),
                              geom.x(),
                              geom.y(),
                              font.pixelSize())
            self.arr.append(data)
        except:
            print("widget has no geometry")
    
    #
    # qrect is the geometry of the parent widget
    # this is called when the geometry changes
    #
    def setGeometry(self,qrect):
        for data in self.arr:
            widget=data.widget
            widget.setGeometry(self.reScale(data,qrect))
            if data.fontSize != -1:
                font = widget.font()
                if data.fontSize >= 1:
                    font.setPixelSize(data.fontSize)
                    widget.setFont(font)
        self.oldParentSize.setWidth(float(qrect.width()))
        self.oldParentSize.setHeight(float(qrect.height()))

    #
    # resize and re-position the widget
    #
    def reScale(self,data,parentRect):
        wScale = float(parentRect.width()) / self.oldParentSize.width()
        hScale = float(parentRect.height()) / self.oldParentSize.height()
        w    = data.w  * wScale
        h    = data.h  * hScale
        x    = data.x  * wScale
        y    = data.y  * hScale
        data.w = w
        data.h = h
        data.x = x
        data.y = y
        if data.fontSize != -1:
            data.fontSize = hScale * data.fontSize
            # fontsize can lose precision, restore the original value
            # here if the width is within 20 pixels
            if parentRect.width() > data.originalWidth - 10 and parentRect.width() < data.originalWidth + 10:
                data.fontSize = data.originalFontSize
        return QtCore.QRect(x,y,w,h)

# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
# For Qt Group widgets
# something to hold the scaled layout and the group widget
@dataclass
class Group:
    scaledLayout : ScaledLayout
    group        : QtWidgets.QGroupBox

# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
# The item's width, height, and position must be
# preserved as floating point values. The normal
# geometry uses ints. We will lose resolution
# with every resize if we use that, so save as floats.
#
class WidgetData():
    def __init__(self,name, item, width, height, xpos, ypos, fontSize):
        self.name = name
        self.widget = item
        self.w = float(width)
        self.h = float(height)
        self.x = xpos
        self.y = ypos
        self.fontSize = float(fontSize)
        self.originalFontSize = float(fontSize)
        parentWidget = item.parent()
        self.originalWidth = parentWidget.width()


# ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
# This is the container widget that will resize all child widgets
class PlacerWidget(QtWidgets.QWidget):

    def __init__(self,parent = None):
       super(PlacerWidget, self).__init__(parent)
       self.scaledLayout = None
       self.fontSize = 0
       self.groups = []

    #
    # only widgets have geometry
    #
    def isWidget(self,widget):
        isA = True
        try:
            geom = widget.geometry()
        except:
            isA = False
        return isA

    #
    # To add a group widget, we need to add all the children
    # of the group
    #
    def addGroup(self,group,parentWidget = None):
        kids = group.children()
        if len(kids) > 0:
            if parentWidget == None:
                parent = group
            else:
                parent = parentWidget
            scaledLayout = ScaledLayout(parent)
            self.addChildren(group,scaledLayout)
            g = Group(scaledLayout,parent)
            self.groups.append(g)

    #
    # add all child widgets to the scaled layout
    #
    def addChildren(self,parent,scaledLayout):
        list = parent.children()
        for w in list:
            if self.isWidget(w):
                name = w.__class__.__name__
                if  name == 'QGroupBox' or name == 'QFrame':
                    self.addGroup(w)
                elif name == 'QTabWidget':
                    for i in range(w.count()):
                        tab = w.widget(i)
                        self.addGroup(tab,w)
                self.adjustFont(w)
                scaledLayout.addWidget(w)
    #
    # You can set the font size of every widget by calling
    # this function.  Every font added will then have the
    # specfied font size, and the font will be resizeable.
    # Setting it back to 0 will cancel the function.
    #
    def setFontSize(self,size):
        self.fontSize = size

    def adjustFont(self,widget):
        if self.fontSize > 0:
            font = widget.font()
            font.setPixelSize(self.fontSize)
            widget.setFont(font)

    #
    # Call delete only after "show" has been called.
    # Otherwise there is no scaledLayout
    #
    def deleteChild(self,w):
        # remove the widget from the layout
        self.scaledLayout.deleteWidget(w)
        # delete the widget
        w.deleteLater()

    #
    # We create a new scaled layout the first time a
    # resize() is called. All the child widgets should be
    # present at this time.
    #
    def newScaledLayout(self):
        if self.scaledLayout == None:
            self.scaledLayout = ScaledLayout(self)
            self.addChildren(self,self.scaledLayout)

    def resizeGroups(self):
        for g in self.groups:
            layout = g.scaledLayout
            geom   = g.group.geometry()
            layout.setGeometry(geom)

    def resizeEvent(self,event):
        self.newScaledLayout()
        self.scaledLayout.setGeometry(self.rect())
        self.resizeGroups()

# simple test
if __name__ == '__main__':
    import sys
    app = QtWidgets.QApplication(sys.argv)
    w = PlacerWidget(None)
    w.setFontSize(12)
    w.resize(600,400)

    b = QtWidgets.QPushButton("BUTTON 1",w)
    b.move(40,40)
    b.resize(200,100)
            
    b = QtWidgets.QPushButton("BUTTON 2",w);
    b.move(300,40)
    b.resize(200,100)

    # test if delete works
    b = QtWidgets.QPushButton("BUTTON 3",w);
    b.move(40,200)
    b.resize(200,100)

    w.show()
    w.deleteChild(b)         # only after show!
    sys.exit(app.exec_())

File: ruby/placerWidget.rb
require 'Qt'

class ScaledLayout

    #
    # The item's width, height, and position must be
    # preserved as floating point values. The normal
    # geometry uses ints, and we will lose resolution
    # with every resize if we use that.
    #
    class WidgetData
        attr_accessor :name, :widget, :w, :h, :x, :y, :fontSize,
                      :originalFontSize, :originalWidth
        
        def initialize(name, item, width, height, xpos, ypos, fontSize)
            @name = name
            @widget = item
            @w = width.to_f
            @h = height.to_f
            @x = xpos
            @y = ypos
            @fontSize = fontSize.to_f
            @originalFontSize = fontSize.to_f
            parentWidget = item.parent
            @originalWidth = parentWidget.width
        end
    end

    def initialize(parent)      # parent required!
        @arr = Array.new
        geom = parent.geometry
        @oldParentSize = Qt::SizeF.new(geom.width,geom.height)
    end

    def deleteWidget widget
        for i in 0...@arr.size
            data = @arr[i]
            if widget == data.widget
                @arr.delete_at(i)
                break
            end
        end
    end
    
    def addWidget(widget)
        name = "unknown"
        begin
            if widget.objectName
                name = widget.objectName
            end
        rescue
            puts 'widget has no object name'
        end
        
        #
        # some children of widgets are not "widgets" and won't
        # have any geometry. The 'rescue' will take care of
        # that
        #
        begin
            geom = widget.geometry
            font = widget.font
            data = WidgetData.new( name,
                                   widget,
                                   geom.width.to_f,
                                   geom.height.to_f,
                                   geom.x.to_f,
                                   geom.y.to_f,
                                   font.pixelSize)
            @arr << data
        rescue
            puts "widget has no geometry #{widget.objectName}"
        end
    end
    
    #
    # qrect is the geometry of the parent widget
    # this is called when the geometry changes
    #
    def setGeometry(qrect)      # virtual, but needed
        for i in 0...@arr.size
            data = @arr[i]
            widget = data.widget
            widget.setGeometry reScale(data,qrect)

            if data.fontSize != -1 
                font = widget.font
                if data.fontSize >= 1
                    font.setPixelSize data.fontSize
                    widget.setFont font
                end
            end
        end
        @oldParentSize.width  = qrect.width.to_f
        @oldParentSize.height = qrect.height.to_f
    end


    # resize and re-position the widget
    def reScale(data,parentRect)
        wScale = parentRect.width.to_f  / @oldParentSize.width
        hScale = parentRect.height.to_f / @oldParentSize.height
        w    = data.w  * wScale
        h    = data.h  * hScale
        x    = data.x  * wScale
        y    = data.y  * hScale
        data.w = w
        data.h = h
        data.x = x
        data.y = y
        if data.fontSize != -1
            data.fontSize = hScale * data.fontSize
            # fontsize can lose precision, restore the original value here if the width is
            # within 20 pixels
            if parentRect.width > data.originalWidth - 10 && parentRect.width < data.originalWidth + 10
                data.fontSize = data.originalFontSize
            end
        end
        return Qt::Rect.new(x,y,w,h)
    end
end


class PlacerWidget < Qt::Widget

    GroupStruct = Struct.new(:group,:scaledLayout)
    
    def initialize (parent = nil)
        super parent
        @scaledLayout = nil
        @fontSize = 0
        @groups = Array.new
    end

    #
    # only widgets have geometry
    #
    def isWidget?(widget)
        isA = true
        begin
            geom = widget.geometry
        rescue
            isA = false
        end
        return isA
    end

    def addGroup(group,parentWidget = nil)

        if group.children.size > 0
            parent = parentWidget == nil ? group : parentWidget
            scaledLayout = ScaledLayout.new parent
            addChildren(group,scaledLayout)
            g = GroupStruct.new parent,scaledLayout
            @groups << g
        end
    end

    #
    # add all child widgets to the scaled layout
    #
    def addChildren(parent,scaledLayout)

        list = parent.children
        list.each do |w|
            if isWidget? w
                if w.className == 'Qt::GroupBox' ||
                   w.className == 'Qt::Frame'
                    addGroup(w)
                elsif w.className == 'Qt::TabWidget'
                    (0...w.count).each do |i|
                        tab = w.widget i
                        addGroup(tab,w)
                    end
                end
                adjustFont w
                scaledLayout.addWidget w
            end
        end
    end

    #
    # You can set the font size of every widget by calling
    # this functions.  Every font added will then have the
    # specfied font size, and the font will be resizeable.
    # Setting it back to 0 will cancel the function.
    #
    def setFontSize size
        @fontSize = size
    end

    def adjustFont(widget)
        if @fontSize > 0
            font = widget.font
            font.setPixelSize @fontSize
            widget.setFont font
        end
    end

    #
    # Call delete only after "show" has been called.
    # Otherwise there is no scaledLayout
    #
    def deleteChild w
        @scaledLayout.deleteWidget w # remove from the scaled layout
        w.deleteLater                # delete the widget
    end
    
    #
    # we create a new scaled layout the
    # first time that resize was called
    # All the child widgets should be
    # present
    #
    def newScaledLayout
        if @scaledLayout == nil
            @scaledLayout = ScaledLayout.new self
            addChildren self,@scaledLayout
        end
    end

    def resizeGroups
        @groups.each do |g|
            g.scaledLayout.setGeometry(g.group.geometry)
        end
    end

    def resizeEvent(event)
        newScaledLayout
        @scaledLayout.setGeometry rect
        resizeGroups
    end

end

if __FILE__ == $0
    Qt::Application.new(ARGV) do
        w = PlacerWidget.new
        w.resize 600,400

        b = Qt::PushButton.new("BUTTON 1",w);
        b.move(40,40)
        b.resize(200,100)
            
        b = Qt::PushButton.new("BUTTON 2",w);
        b.move(300,40)
        b.resize(200,100)
        b2  = b
        
        # test if delete works

        b = Qt::PushButton.new("BUTTON 3",w);
        b.move(40,200)
        b.resize(200,100)

        w.show
        w.deleteChild b         # only after show!

        exec
        exit 0
    end
end
